docs/projects/insertion-sort/_posts/2019-01-22-python.md

```diff
@@ -0,0 +1,218 @@
+empty to begin with.
+
+We loop through every element in the unsorted list `for x in xs`. Then we call the `insert()`
+function to add `x` to the `new_xs` (the sorted list) in the correct position.
+Each time the `insert()` function is called it returns a sorted list which we assign
+to `new_xs`. Finally, when we've looped through every item in `xs`, we return the sorted list
+`new_xs` which will then get printed on the terminal to the user.
+
+Taking a look at an example where `xs = [5, 3, 10]`
+
+1st
+
+* `x = 5`
+* `insert(5, [])`
+* `new_xs = [5]`
+
+2nd
+
+* `x = 3`
+* `insert(3, [5])`
+* `new_xs = [3, 5]`
+
+3rd
+
+* `x = 10`
+* `insert(10, [3, 5])`
+* `new_xs = [3, 5, 10]`
+
+### Insert
+
+```python
+def insert(x, xs):
+    left = list(takewhile(lambda i: i < x, xs))
+    right = xs[len(left):]
+    return left + [x] + right
+```
+
+This function takes two parameters `x`, which is an element from our unsorted list, and
+`xs`, which is the list to add `x` to such that `xs` remains sorted.
+A High-level overview of this function is that the `left` variable will be a list that stores
+all elements less than `x` from `xs` and `right` will store all elements greater than (or equal)
+to `x`. That way we can "insert" `x` between these two lists. Hence the return statement
+looking like `return left + [x] + right` and this returned list will therefore be sorted.
+
+Let's take a look at the first line `left = list(takewhile(lambda i: i < x, xs))` it looks a bit
+complicated so let's break it down. The first part `lambda i: i < x, xs` is a lambda function
+which is a small anonymous unnamed function.
+In this case `i` is an element of `xs` and we want all `i`'s less than `x`.
+
+Then `takewhile(lambda i: i < x, xs)` where `takewhile()` takes a predicate (our lambda function) and a list
+It stops iterating over our list as soon as the lambda function evaluates to False. It then returns
+all the elements up to that index. Lets take a look at an example where `xs = [4, 7, 10]` and `x = 8`.
+The first two items of `xs` would evaluate as True since 4 and 7 are
+less than 8, so takewhile would store 4 and 7 but not 10.
+The `takewhile()` function returns a `takewhile` object but we want a list so we convert that into a
+list hence `list(takewhile(lambda i: i < x, xs))`. So the `left` variable will store all numbers
+less than `x`, since `xs` is already sorted.
+
+Moving onto `right = xs[len(left):]`, `len(left)` returns the length of the left list.
+Then we do some index splicing, you can learn more about it
+[here](https://www.pythoncentral.io/how-to-slice-listsarrays-and-tuples-in-python/).
+Index splicing is used to get part of a list in Python. In this case we getting every element in
+the list that's not already in `left`. We can do this because we know `xs` is already sorted.
+If `xs = [1, 4, 6, 8]` and `x = 7` then, `left = [1, 4, 6]` (all elements < 7). So `len(left) = 3`,
```

Comment:
> So `len(left) = 3`,
> so `right = xs[3:]`, where `[3:]` gets all elements from `xs` not including the first `3` elements,
> therefore `right = 8`.

You'll have to help me out with this one! There are four sections here separated by commas, and I don't know how to reword it to make it more clear.